pkg/util/httprequest/httprequest.go

```diff
@@ -87,11 +78,26 @@ func SchemeHost(req *http.Request) (string /*scheme*/, string /*host*/) {
 		host = req.URL.Host
 	}
 
+	// If both X-Forwarded-Host and X-Forwarded-Port are sent, use the explicit port info
+	if forwardedPort := forwarded("Port"); len(forwardedHost) > 0 && len(forwardedPort) > 0 {
+		if h, _, err := net.SplitHostPort(forwardedHost); err == nil {
+			host = net.JoinHostPort(h, forwardedPort)
```

Comment:
it's not great to reset `host` from forwardedHost and forwardedPort after the switch block where we set `host` from the different sources. would prefer all the logic be contained in the different cases of that block.

```
hasExplicitHost := func(h string) bool {
  _, _, err := net.SplitHostPort(h)
  return err == nil
}

host := ""
hostHadExplicitPort := false
switch {
case len(forwardedHost) > 0:
	host = forwardedHost
	hostHadExplicitPort = hasExplicitHost(host)

	// If both X-Forwarded-Host and X-Forwarded-Port are sent, use the explicit port info
	if forwardedPort := forwarded("Port"); len(forwardedPort) > 0 {
		if h, _, err := net.SplitHostPort(forwardedHost); err == nil {
			host = net.JoinHostPort(h, forwardedPort)
		} else {
			host = net.JoinHostPort(forwardedHost, forwardedPort)
		}
	}

case len(req.Host) > 0:
	host = req.Host
	hostHadExplicitPort = hasExplicitHost(host)

case len(req.URL.Host) > 0:
	host = req.URL.Host
	hostHadExplicitPort = hasExplicitHost(host)
}
```